app/com/linkedin/drelephant/spark/heuristics/StagesAnalyzer.scala

```diff
@@ -0,0 +1,498 @@
+      stageData.tasks.map(tasks => (stageData.stageId, tasks.values))
+    }.toMap
+    
+    val failedTasks = failedTasksStageMap.get(stageId)
+
+    val details = new ArrayBuffer[String]()
+
+    val taskFailureSeverity = taskFailureRateSeverityThresholds.severityOf(
+      stageData.numFailedTasks.toDouble / stageData.numTasks)
+    if (hasSignificantSeverity(taskFailureSeverity)) {
+      details += s"Stage $stageId has ${stageData.numFailedTasks} failed tasks."
+    }
+
+    val score = Utils.getHeuristicScore(taskFailureSeverity, stageData.numFailedTasks)
+
+    val (numTasksWithOOM, oomSeverity) =
+      checkForSpecificTaskError(stageId, stageData, failedTasks,
+        StagesWithFailedTasksHeuristic.OOM_ERROR, "of OutOfMemory exception.",
+        details)
+
+    val (numTasksWithContainerKilled, containerKilledSeverity) =
+      checkForSpecificTaskError(stageId, stageData, failedTasks,
+        StagesWithFailedTasksHeuristic.OVERHEAD_MEMORY_ERROR,
+        "the container was killed by YARN for exeeding memory limits.", details)
+
+    TaskFailureResult(taskFailureSeverity, score, details, oomSeverity, containerKilledSeverity,
+      stageData.numFailedTasks, numTasksWithOOM, numTasksWithContainerKilled)
+  }
+
+  /**
+    * Check the stage for a high ratio of time spent in GC compared to task run time.
+    *
+    * @param stageId stage ID.
+    * @param stageData stage data.
+    * @return result of GC analysis for the stage.
+    */
+  private def checkForGC(stageId: Int, stageData: StageData): SimpleStageAnalysisResult = {
+    var gcTime = 0.0D
+    var taskTime = 0.0D
+    val severity = stageData.taskSummary.map { task =>
+      gcTime = task.jvmGcTime(DISTRIBUTION_MEDIAN_IDX)
+      taskTime = task.executorRunTime(DISTRIBUTION_MEDIAN_IDX)
+      DEFAULT_GC_SEVERITY_A_THRESHOLDS.severityOf(gcTime / taskTime)
+    }.getOrElse(Severity.NONE)
+
+    val score = Utils.getHeuristicScore(severity, stageData.numTasks)
+
+    val details = if (hasSignificantSeverity(severity)) {
+      Seq(s"Stage ${stageId}: tasks are spending signficant time in GC (median task GC time is " +
+        s"${Utils.getDuration(gcTime.toLong)}, median task runtime is " +
+        s"${Utils.getDuration(gcTime.toLong)}")
+    } else {
+      Seq.empty
+    }
+
+    new SimpleStageAnalysisResult(severity, score, details)
+  }
+
+  /**
+    * Check the stage for tasks that failed for a specified error.
+    *
+    * @param stageId stage ID.
+    * @param stageData stage data.
+    * @param failedTasks list of failed tasks.
+    * @param taskError the error to check for.
+    * @param errorMessage the message/explanation to print if the the specified error is found.
+    * @param details information and recommendations -- any new recommendations
+    *                from analyzing the stage for errors causing tasks to fail will be appended.
+    * @return
+    */
+  private def checkForSpecificTaskError(
```

Comment:
Whats the use of this check ? . If the task failed because of some reason other than OOM (which is handled for checkForTaskFailure) and the job passes (because of task retry) , why user should bother about the reason of task failed .?

Reply:
It's also used to identify container killed by YARN for exeeding memory limits.